### ISSUE TITLE
Computed field is recomputed when unrelated M2M is affected

### ISSUE THREAD

**Issue by mobiware** [netzkolchose/django-computedfields#43]
A model has a computed field depending on an M2M on a related model. This related model has another M2M as well. This second M2M is not part of any Computed Field dependency.

Yet, when this second M2M is modified (an object is added for instance), the model with computed field gets updated even though it does not depend on the modified M2M.

Here's a repro case:
```
class Group(models.Model):
    pass


class Item(models.Model):
    pass


class User(models.Model):
    groups = models.ManyToManyField(Group, related_name="users")
    items = models.ManyToManyField(Item, related_name="users")


class Agent(ComputedFieldsModel):
    user = models.OneToOneField(User, related_name="agent", on_delete=models.CASCADE)

    @computed(models.IntegerField(default=0), depends=[["user.items", ["id"]]])
    def counter(self):
        # This is used to detect when Agent gets updated
        return self.counter + 1


def test_alter_non_dependent_m2m(self):
    group = Group.objects.create()
    user = User.objects.create()
    agent = Agent.objects.create(user=user)

    # Agent should have been saved only once
    # This assert passes
    self.assertEqual(agent.counter, 1)

    # Note that we have no dependency on "groups" anywhere
    user.groups.add(group)

    agent.refresh_from_db()

    # Yet, this last test fails, we find that agent.counter == 2, which means
    # that "agent" was updated when we added a group
    self.assertEqual(agent.counter, 1)
```

**Comment by mobiware** [comment]
I see that `m2m_handler` is registered to the `m2m_changed` signal unconditionally (for all models), isn't it possible to only register that handler to `sender`s that are involved in computed fields dependencies ?

**Comment by jerch** [comment]
@mobiware Yes, I am aware of that problem. It is based on the fact, the Django's `m2m_changed` signal gives not enough meta information about the affected m2m relation. Thus the code currently triggers updates on all relations in all directions. m2m relations are really bad, not only because if this.

If you have an idea, how to overcome this, I am all ears. The code causing this is the `m2m_changed` handler in `handlers.py`.

**Comment by jerch** [comment]
@mobiware The problem might be solvable, if the resolver would always descent into the `through` model, even the hidden default one, and not using the m2m signals anymore. This way the resolver can just do its fk/fk_reverse magic. But I am not sure yet, if the resolver can catch all instance actions that way. Needs investigation.

**Comment by jerch** [comment]
@mobiware Lol forget what I wrote above. Looked in the django source and started to wonder, why the `m2m_changed` signal is triggered on the through model. Turns out - sender argument it the through model, haha. For some reason I overlooked this. Even worse - I already use it in the `preclear` branch, but the other branches ignore it. So basically the handler code is insufficient.

**Comment by jerch** [comment]
@mobiware Your example above is somewhat linked to the `update_fields` handling by the resolver - whenever `update_dependent(some_queryset, update_fields=None)` is called without explicit fields (`None` is default), it will trigger a full update on the first dependency level. This behavior is needed to provide a forced update possibility, otherwise desync state cannot be repaired easily anymore.

For m2m relations the handler triggers `update_dependent` into both directions with no update_fields. Applied to your example above `update_dependent(Group.objects.filter(pk__in=group_pks), update_fields=None)` and `update_dependent(user_instance, update_fields=None)` are executed. Now through the implicit dependency expansion of fk fields, `Agent.counter` also depends on user changes (through your o2o relation), thus will be triggered by that call.

The issue boils down to the question, if we can find a way to apply `update_fields` at the m2m handler stage to narrow down follow-up updates. But this does not work with m2m relations directly, as update fields must be concrete fields (which is the reason, why the right side in `depends` can only contain concrete fields). It would work with the fk fields on the `through` model though. But to get this working the resolver has to give up the easy m2m filtering on m2m relations, and always go with the through model, which bloats the dependency tree alot. Not sure yet, if it is worth to do that (it introduces tons of problems at the queryset construction stage, not even sure how to join on the hidden through table there).

**Comment by jerch** [comment]
@mobiware Did some more in-depth analysis. This is what the resolver currently comes up with (resolver map):

```python
{
    <class 'models.User'>: {
        'items': {
            <class 'models.Agent'>: ({'counter'}, {'user'})
        }
    },
    <class 'models.Item'>: {
        'id': {
            <class 'models.Agent'>: ({'counter'}, {'user__items'})
        }
    }
}
```

The map reads as follows:
- top level keys are models, that where changed (instance/qs given to `update_dependent`)
- subkeys are fields on the input model cfs on foreign models depend on
- subsubkeys are affected models with cfs
- values are sets of cfs and querystrings to pull the new data for update

So basically the resolver knows 2 update rules:
- `User.items` updates `Agent.counter` via `Agent.objects.filter(user__in=altered_users)`
- `Item.id` updates `Agent.counter` via `Agent.objects.filter(user__items__in=altered_items)`

Note that there is no update rule for `User.groups`. It still gets caught by the m2m handler, since it cannot derive the changed relation field atm, thus triggering an ALL rule running all updates listed for `User`.

To fix the ALL rule handling, two things have to be done:
- correctly derive/pull affected m2m relation field in handler
- rework `_querysets_for_update` so it can handle non concrete fields in `update_fields` for the m2m case (migth be a special case in the end)

**Comment by mobiware** [comment]
@jerch is this map built by `_querysets_for_update` ? Perhaps we should call `_querysets_for_update` in the M2M handler and then walk through that graph to see if `sender` (the `through` model) or `model` (the model of the object that was added to a relationship, here it would be `Group`) are present in the graph. If neither is present, then it means the M2M change doesn't affect any computed field on `instance`

Of course we'd have to also take into account the `reverse` case (i.e. if I do `group.users.add(user)` instead of the `user.groups.add(group)` in my test case) and the other M2M events (remove, clear).

Regarding your comment about non concrete fields handling, do you mean that there's another issue with M2M specific to non concrete fields ? Is it when the computed fields dependent on an M2M relation is non concrete, or when the M2M field is non concrete ?

**Comment by jerch** [comment]
> is this map built by `_querysets_for_update`?

Nope, its the static resolver map built during startup. `_querysets_for_update` uses it to construct the live update querysets for a single `update_dependent` call. There is no graph handling involved at runtime resolving.

> Of course we'd have to also take into account the reverse case (i.e. if I do group.users.add(user) instead of the user.groups.add(group) in my test case) and the other M2M events (remove, clear).

Yes the whole handler for signal types has to be patched with the updated m2m field notion. The reverse case is quite simple, the handler has to trigger in both directions anyway, thus the reverse case just the flipped case.

> Regarding your comment about non concrete fields handling, do you mean that there's another issue with M2M specific to non concrete fields ? Is it when the computed fields dependent on an M2M relation is non concrete, or when the M2M field is non concrete ?

`update_fields` is currently a hybrid, it is used to denote real local concrete fields (used to update during `save` and by `bulk_updater`). On the higher inter-model level it is used to denote any field deps, but for fk fields those happen to be concrete as well. M2M fields are never concrete in that sense, so this might need some special handling during qs construction.

**Comment by jerch** [comment]
@mobiware Plz see #45.

**Comment by mobiware** [comment]
@jerch sorry I couldn't review before you merged, thanks a lot for addressing this !

**Comment by jerch** [comment]
@mobiware All good, I did some extensive testing, also the code looks more uniform now. It also revealed the fact, that m2m fields were not auto-expanded yet. Imho a good find after all. :+1:
